# One Hot encoding for features with lots of categories

In [8]:
import numpy as np
import pandas as pd
data=pd.read_csv('train.csv',usecols=['X1','X2','X3','X4','X5','X6'])
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [9]:
data.columns.unique()

Index(['X1', 'X2', 'X3', 'X4', 'X5', 'X6'], dtype='object')

In [19]:
data.columns.value_counts()

X1    1
X3    1
X6    1
X4    1
X2    1
X5    1
dtype: int64

In [13]:
for col in data.columns:
    print(col,' :',len(data[col].unique()),'labels')

X1  : 27 labels
X2  : 44 labels
X3  : 7 labels
X4  : 4 labels
X5  : 29 labels
X6  : 12 labels


In [14]:
pd.get_dummies(data,drop_first=True).shape

(4209, 117)

for variable of 6 we get 117 additional new features

In [42]:
# top 10 most frequent categories for variable x2

data.X2.value_counts().sort_values(ascending=False).head(10)


as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
Name: X2, dtype: int64

In [23]:
top10=[x for x in data.X2.value_counts().sort_values(ascending=False).head(10).index]
top10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

encoding top10 most frequent categories for x2 and appending them to the table

In [25]:
for label in top10:
    data[label]=np.where(data['X2']==label,1,0)
    
    

In [26]:
data

X1  X2 X3 X4  X5 X6  as  ae  ai  m  ak  r  n  s  f  e
0     v  at  a  d   u  j   0   0   0  0   0  0  0  0  0  0
1     t  av  e  d   y  l   0   0   0  0   0  0  0  0  0  0
2     w   n  c  d   x  j   0   0   0  0   0  0  1  0  0  0
3     t   n  f  d   x  l   0   0   0  0   0  0  1  0  0  0
4     v   n  f  d   h  d   0   0   0  0   0  0  1  0  0  0
...  ..  .. .. ..  .. ..  ..  ..  .. ..  .. .. .. .. .. ..
4204  s  as  c  d  aa  d   1   0   0  0   0  0  0  0  0  0
4205  o   t  d  d  aa  h   0   0   0  0   0  0  0  0  0  0
4206  v   r  a  d  aa  g   0   0   0  0   0  1  0  0  0  0
4207  r   e  f  d  aa  l   0   0   0  0   0  0  0  0  0  1
4208  r  ae  c  d  aa  g   0   1   0  0   0  0  0  0  0  0

[4209 rows x 16 columns]

In [28]:
def one_hot_top(data,label,top):
    
    for x in top:
        data[label+'_'+x]=np.where(data[label]==x,1,0)
    
    
data=pd.read_csv('train.csv',usecols=['X1','X2','X3','X4','X5','X6'])
one_hot_top(data,'X2',top10)

Encoding all the categorical variables one by one and appending them to the dataset

In [29]:
data

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0     v  at  a  d   u  j      0      0      0     0      0     0     0     0   
1     t  av  e  d   y  l      0      0      0     0      0     0     0     0   
2     w   n  c  d   x  j      0      0      0     0      0     0     1     0   
3     t   n  f  d   x  l      0      0      0     0      0     0     1     0   
4     v   n  f  d   h  d      0      0      0     0      0     0     1     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...    ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0      0     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0      0     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0      0     1     0     0   
4207  r   e  f  d  aa  l      0      0      0     0      0     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0      0     0     0     0   

      X2_f  X2_e  
0        0     0  
1        0     0  
2        0     0  
3        0     0  
4        0     0  
...    ...   ...  
4204     0     0  
4205     0     0  
4206     0     0  
4207     0     1  
4208     0     0  

[4209 rows x 16 columns]

In [34]:
top10x1=[x for x in data.X1.value_counts().sort_values(ascending=False).head(10).index]
top10x1

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [35]:
one_hot_top(data,'X1',top10x1)

In [36]:
data

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_aa  X1_s  X1_b  \
0     v  at  a  d   u  j      0      0      0     0  ...      0     0     0   
1     t  av  e  d   y  l      0      0      0     0  ...      0     0     0   
2     w   n  c  d   x  j      0      0      0     0  ...      0     0     0   
3     t   n  f  d   x  l      0      0      0     0  ...      0     0     0   
4     v   n  f  d   h  d      0      0      0     0  ...      0     0     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...  ...    ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0  ...      0     1     0   
4205  o   t  d  d  aa  h      0      0      0     0  ...      0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0  ...      0     0     0   
4207  r   e  f  d  aa  l      0      0      0     0  ...      0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0  ...      0     0     0   

      X1_l  X1_v  X1_r  X1_i  X1_a  X1_c  X1_o  
0        0     1     0     0     0     0     0  
1        0     0     0     0     0     0     0  
2        0     0     0     0     0     0     0  
3        0     0     0     0     0     0     0  
4        0     1     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0     0  
4205     0     0     0     0     0     0     1  
4206     0     1     0     0     0     0     0  
4207     0     0     1     0     0     0     0  
4208     0     0     1     0     0     0     0  

[4209 rows x 26 columns]

In [37]:
top10x3=[x for x in data.X3.value_counts().sort_values(ascending=False).head(10).index]
one_hot_top(data,'X3',top10x3)
data

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_a  X1_c  X1_o  \
0     v  at  a  d   u  j      0      0      0     0  ...     0     0     0   
1     t  av  e  d   y  l      0      0      0     0  ...     0     0     0   
2     w   n  c  d   x  j      0      0      0     0  ...     0     0     0   
3     t   n  f  d   x  l      0      0      0     0  ...     0     0     0   
4     v   n  f  d   h  d      0      0      0     0  ...     0     0     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...  ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0  ...     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0  ...     0     0     1   
4206  v   r  a  d  aa  g      0      0      0     0  ...     0     0     0   
4207  r   e  f  d  aa  l      0      0      0     0  ...     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0  ...     0     0     0   

      X3_c  X3_f  X3_a  X3_d  X3_g  X3_e  X3_b  
0        0     0     1     0     0     0     0  
1        0     0     0     0     0     1     0  
2        1     0     0     0     0     0     0  
3        0     1     0     0     0     0     0  
4        0     1     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...  
4204     1     0     0     0     0     0     0  
4205     0     0     0     1     0     0     0  
4206     0     0     1     0     0     0     0  
4207     0     1     0     0     0     0     0  
4208     1     0     0     0     0     0     0  

[4209 rows x 33 columns]

In [38]:
top10x3

['c', 'f', 'a', 'd', 'g', 'e', 'b']

In [39]:
top10x4=[x for x in data.X4.value_counts().sort_values(ascending=False).head(10).index]
one_hot_top(data,'X4',top10x4)
data

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X3_f  X3_a  X3_d  \
0     v  at  a  d   u  j      0      0      0     0  ...     0     1     0   
1     t  av  e  d   y  l      0      0      0     0  ...     0     0     0   
2     w   n  c  d   x  j      0      0      0     0  ...     0     0     0   
3     t   n  f  d   x  l      0      0      0     0  ...     1     0     0   
4     v   n  f  d   h  d      0      0      0     0  ...     1     0     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...  ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0  ...     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0  ...     0     0     1   
4206  v   r  a  d  aa  g      0      0      0     0  ...     0     1     0   
4207  r   e  f  d  aa  l      0      0      0     0  ...     1     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0  ...     0     0     0   

      X3_g  X3_e  X3_b  X4_d  X4_a  X4_c  X4_b  
0        0     0     0     1     0     0     0  
1        0     1     0     1     0     0     0  
2        0     0     0     1     0     0     0  
3        0     0     0     1     0     0     0  
4        0     0     0     1     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...  
4204     0     0     0     1     0     0     0  
4205     0     0     0     1     0     0     0  
4206     0     0     0     1     0     0     0  
4207     0     0     0     1     0     0     0  
4208     0     0     0     1     0     0     0  

[4209 rows x 37 columns]

In [40]:
top10x5=[x for x in data.X5.value_counts().sort_values(ascending=False).head(10).index]
one_hot_top(data,'X5',top10x5)
data

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X5_w  X5_v  X5_q  \
0     v  at  a  d   u  j      0      0      0     0  ...     0     0     0   
1     t  av  e  d   y  l      0      0      0     0  ...     0     0     0   
2     w   n  c  d   x  j      0      0      0     0  ...     0     0     0   
3     t   n  f  d   x  l      0      0      0     0  ...     0     0     0   
4     v   n  f  d   h  d      0      0      0     0  ...     0     0     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...  ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0  ...     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0  ...     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0  ...     0     0     0   
4207  r   e  f  d  aa  l      0      0      0     0  ...     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0  ...     0     0     0   

      X5_r  X5_d  X5_s  X5_n  X5_m  X5_p  X5_i  
0        0     0     0     0     0     0     0  
1        0     0     0     0     0     0     0  
2        0     0     0     0     0     0     0  
3        0     0     0     0     0     0     0  
4        0     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0     0  
4205     0     0     0     0     0     0     0  
4206     0     0     0     0     0     0     0  
4207     0     0     0     0     0     0     0  
4208     0     0     0     0     0     0     0  

[4209 rows x 47 columns]

In [41]:
top10x6=[x for x in data.X6.value_counts().sort_values(ascending=False).head(10).index]
one_hot_top(data,'X6',top10x6)
data

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X6_g  X6_j  X6_d  \
0     v  at  a  d   u  j      0      0      0     0  ...     0     1     0   
1     t  av  e  d   y  l      0      0      0     0  ...     0     0     0   
2     w   n  c  d   x  j      0      0      0     0  ...     0     1     0   
3     t   n  f  d   x  l      0      0      0     0  ...     0     0     0   
4     v   n  f  d   h  d      0      0      0     0  ...     0     0     1   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...  ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0  ...     0     0     1   
4205  o   t  d  d  aa  h      0      0      0     0  ...     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0  ...     1     0     0   
4207  r   e  f  d  aa  l      0      0      0     0  ...     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0  ...     1     0     0   

      X6_i  X6_l  X6_a  X6_h  X6_k  X6_c  X6_b  
0        0     0     0     0     0     0     0  
1        0     1     0     0     0     0     0  
2        0     0     0     0     0     0     0  
3        0     1     0     0     0     0     0  
4        0     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0     0  
4205     0     0     0     1     0     0     0  
4206     0     0     0     0     0     0     0  
4207     0     1     0     0     0     0     0  
4208     0     0     0     0     0     0     0  

[4209 rows x 57 columns]